Update shape of two-way binding instructions

crisbeto opened the pull request:
Includes some changes to update the instructions generated for two-way bindings from:

```js
function template(rf, ctx) {
  if (rf & 1) {
    listener("valueChange", function($event) {
      return ctx.name = $event;
    });
  } 
  if (rf & 2) {
    property("value", ctx.name);
  }
}
```

To:

```js
function template(rf, ctx) {
  if (rf & 1) {
    twoWayListener("valueChange", function($event) {
      twoWayBindingSet(ctx.name, $event) || (ctx.name = $event);
      return $event;
    });
  } 
  if (rf & 2) {
    twoWayProperty("value", ctx.name);
  }
}
```

The new shape will be necessary for some future work and implementing it required cleaning up some old parsing code. There are more details in the individual commits.

### refactor(compiler): implement two-way property instruction
Reworks the compiler so that it generates a `twoWayProperty` instruction, instead of `property`, for the property side of a two-way binding. Currently the new instruction passes through to `property`, but it'll have some two-way-binding-specific logic in subsequent PRs.

### refactor(compiler): update access of members in expression parser
Currently all the members of `_ParseAST` are public, even though they're all used only within the class. This change marks them as private so that it's explicit which ones are intended to be used outside the class.

### refactor(core): introduce two-way listener instructions
Adds the following new instructions:
* `twoWayBindingSet` - used to assign values inside of the listener side of a two-way binding. Currently a noop, but will come into play later.
* `twoWayListener` - used to bind a two-way listener. Currently calls directly into `listener`, but it may be useful in the future.

### refactor(compiler): preserve expression in two-way listeners
Currently the listener side two-way listeners are parsed by appending `=$event` to the raw expression. This is problematic, because:
1. It can interfere with other expressions (see https://github.com/angular/angular/pull/37809).
2. It can lead to confusing error messages because users will see code that they didn't write.
3. It doesn't allow us to further manipulate the expression.

These changes remove the logic that appends `=$event` to resolve the issue. There's also some new logic that checks the expression after it has been parsed to ensure that the result is an assignable expression.

Subsequent commits will update the code that emits the expression to add back the `$event` assignment where it's needed.

### refactor(compiler): update two-way listener emit in definition builder
Updates the template definition builder to emit the new format for the listener side of two-way bindings.

### refactor(compiler): implement new two-way listener shape in pipeline
Implements the new shape of two-way listener instructions in the template pipeline.

### refactor(compiler): allow some invalid expressions in two-way bindings that previously worked by accident
In one of the earlier commits, the logic that appends `=$event` before parsing two-way bindings was removed and some validation was added to prevent unassignable expressions from being used. This ended up being problematic, because previously the parser was incorrectly allowing some invalid expressions which users came to depend on. For example, it transformed `[(value)]="a && a.b"` to `a && (a.b = $event)`.

These changes add some special cases for the common breakages that came up during the TGP.

### refactor(compiler): maintain order between two-way and one-way properties
One of the earlier commits separated one-way and two-way bindings which ended up breaking some internal targets, because it changed the assignment order. These changes bring back the old order.

crisbeto committed:
refactor(compiler): implement two-way property instruction

Reworks the compiler so that it generates a `twoWayProperty` instruction, instead of `property`, for the property side of a two-way binding. Currently the new instruction passes through to `property`, but it'll have some two-way-binding-specific logic in subsequent PRs.

crisbeto committed:
refactor(compiler): update access of members in expression parser

Currently all the members of `_ParseAST` are public, even though they're all used only within the class. This change marks them as private so that it's explicit which ones are intended to be used outside the class.

crisbeto committed:
refactor(core): introduce two-way listener instructions

Adds the following new instructions:
* `twoWayBindingSet` - used to assign values inside of the listener side of a two-way binding. Currently a noop, but will come into play later.
* `twoWayListener` - used to bind a two-way listener. Currently calls directly into `listener`, but it may be useful in the future.

crisbeto committed:
refactor(compiler): preserve expression in two-way listeners

Currently the listener side two-way listeners are parsed by appending `=$event` to the raw expression. This is problematic, because:
1. It can interfere with other expressions (see #37809).
2. It can lead to confusing error messages because users will see code that they didn't write.
3. It doesn't allow us to further manipulate the expression.

These changes remove the logic that appends `=$event` to resolve the issue. There's also some new logic that checks the expression after it has been parsed to ensure that the result is an assignable expression.

Subsequent commits will update the code that emits the expression to add back the `$event` assignment where it's needed.

crisbeto committed:
refactor(compiler): update two-way listener emit in definition builder

Updates the template definition builder to emit the new format for the listener side of two-way bindings.

```js
// Before
listener("ngModelChange", function($event) {
  return ctx.name = $event;
});

// After
ɵɵtwoWayListener("ngModelChange", function($event) {
  ɵɵtwoWayBindingSet(ctx.name, $event) || (ctx.name = $event);
  return $event;
});
```

crisbeto committed:
refactor(compiler): implement new two-way listener shape in pipeline

Implements the new shape of two-way listener instructions in the template pipeline.

crisbeto commented on packages/compiler/src/template_parser/binding_parser.ts:
```diff
@@ -464,26 +463,34 @@ export class BindingParser {
+    // Don't try to validate assignment events if there were other
+    // parsing errors to avoid adding more noise to the error logs.
+    if (isAssignmentEvent && isValid && !this._isAllowedAssignmentEvent(ast)) {
+      this._reportError('Unsupported expression in a two-way binding', sourceSpan);
```
This validation turned out to be quite breaking in the TGP. We'll have to add some special handling for binary expressions. The rest of the PR should be largely unchanged though.

crisbeto committed:
refactor(compiler): allow some invalid expressions in two-way bindings that previously worked by accident

In one of the earlier commits, the logic that appends `=$event` before parsing two-way bindings was removed and some validation was added to prevent unassignable expressions from being used. This ended up being problematic, because previously the parser was incorrectly allowing some invalid expressions which users came to depend on. For example, it transformed `[(value)]="a && a.b"` to `a && (a.b = $event)`.

These changes add some special cases for the common breakages that came up during the TGP.

crisbeto commented on packages/compiler/src/compiler_util/expression_converter.ts:
```diff
@@ -77,6 +53,95 @@ export function convertActionBinding(
+    return wrapAssignmentReadExpression(ast);
+  }
+
+  // However, historically the expression parser was handling two-way events by appending `=$event`
```
I'm not excited by having to introduce this logic, but I don't think there's a better way of handling this in a non-breaking way.

alfaproject replied:
Will you consider adding diagnostics and maybe even deprecate these usages?

crisbeto replied:
I think it's a good idea to have a diagnostic for this but I suspect that completely removing it will be difficult since fixing the expression can change the app's behavior.

crisbeto committed:
refactor(compiler): maintain order between two-way and one-way properties

One of the earlier commits separated one-way and two-way bindings which ended up breaking some internal targets, because it changed the assignment order. These changes bring back the old order.

crisbeto commented on packages/compiler/src/template/pipeline/src/phases/const_collection.ts:
```diff
@@ -98,7 +98,12 @@ const FLYWEIGHT_ARRAY: ReadonlyArray<o.Expression> = Object.freeze<o.Expression[
   private known = new Map<ir.BindingKind, Set<string>>();
-  private byKind = new Map<ir.BindingKind, o.Expression[]>;
+  private byKind = new Map<
+      // Property bindings are excluded here, because they need to be tracked in the same
```
This is a bit less clean than what we had before, but I went it because it's much more explicit about _why_ we need to keep track of one-way and two-way properties together. The `Exclude` here guarantees that we don't try accidentally try to look up `BindingKind.Property` in the map.

Alternatively, this can be fixed by special-casing two-way bindings like this in `add`:

```
add(kind: ir.BindingKind...): void {
  if (kind === ir.BindingKind.TwoWayProperty) {
    kind = ir.BindingKind.Property;
  }
}
```

I decided not to go down this route, because it's a bit of a gotcha and could lead to subtle bugs in the future.

dylhunn replied:
Makes sense to me.

crisbeto commented:
[Passing TGP](http://test/OCL:602972326:BASE:603011338:1706705126537:14f511cd)

dylhunn commented on packages/compiler/src/compiler_util/expression_converter.ts:
```diff
@@ -77,6 +53,43 @@ export function convertActionBinding(
+  // Converts `[(ngModel)]="name"` to `twoWayBindingSet(ctx.name, $event) || (ctx.name = $event)`.
+  convertedAction = new o.ExternalExpr(R3.twoWayBindingSet)
+                        .callFn([convertedAction.expr, EventHandlerVars.event])
+                        .or(convertedAction.expr.set(EventHandlerVars.event))
```
Design question: why can't we just pass `ctx` and always do the assignment inside the instruction, instead of having `||` and falling back to doing the assignment here if it's not a signal?

crisbeto replied:
Do you mean something like `function listener($event) { setValue(ctx, 'someProp', $event) }`. If yes, it's because it'll break with property minification.

dylhunn commented on packages/compiler/src/template/pipeline/ir/src/enums.ts:
```diff
@@ -195,6 +195,11 @@ export enum OpKind {
+   * An operation to bind an expression to the property side of a two-way binding.
+   */
+  TwoWayProperty,
+
```
Adding the new `OpKind` required updates to a couple phases. I think you may have missed a few though:

1. `orderOps`: We should probably have an ordering invariant in order to maximize chaining opportunities. I think that, to maintain the TDB order, you'd probably have to test for *either* `Property` or `TwoWayProperty` kind together. (I think it's actually a bug not to have this: currently, `TwoWayProperty` instructions will *break up* the ordering of existing lists of binding ops, potentially leading to wrong orders compared to TDB, since we "naturally" emit them in different orders.)

And then I'm less sure that you'd need changes to these:

2. `createI18nContexts`: Can this ever result in an extracted i18n attr? I guess that raises the question of whether you can write `[(i18n-foo)]="..."`, which would be an insane thing to do, so maybe it doesn't matter?
3. `naming`: Two-way bindings can never be an animation trigger or have `@`, right?
4. `resolveSanitizers`: can two-way bindings ever have consts requiring sanitization?

crisbeto replied:
1. Done.
2. My understanding is that we don't support i18n in two-way bindings.
3. Nope, they can't be animation triggers.
4. I don't _think_ so since they're always binding into inputs.

dylhunn replied:
Awesome, thanks.

dylhunn commented:
(As mentioned on Slack, this is for the first 4 commits. Will review the next 4 later today. I'd also love to run a TGP with Template Pipeline on.)

crisbeto commented on packages/compiler/src/template/pipeline/src/ingest.ts:
```diff
@@ -1043,15 +1042,22 @@ function createTemplateBinding(
   // If this is a structural template, then several kinds of bindings should not result in an
```
Maybe I'm misunderstanding it, but I think that it should be handled by adding `e.BindingType.TwoWay` to the `BINDING_KINDS` map. This lookup should've taken care of it: https://github.com/angular/angular/blob/main/packages/compiler/src/template/pipeline/src/ingest.ts#L911

dylhunn replied:
Sorry, yes, overlooked that detail -- we are creating a binding op, not an extracted attribute directly.

crisbeto commented on packages/compiler/src/template/pipeline/src/ingest.ts:
```diff
@@ -1043,15 +1042,22 @@ function createTemplateBinding(
+              xref, ir.BindingKind.TwoWayProperty, null, name, null, null, i18nMessage,
+              securityContext);
+      }
     }
```
Yes, `ng-template` can have two-way bindings. I'm not sure which code you're referring to though. Is it the one wrapped in the `if (templateKind === ir.TemplateKind.NgTemplate) {` check right below?

dylhunn reviewed (approved):
OK, remainder looks good to me. Let's also check a Template Pipeline TGP, but then should be good to go.

crisbeto committed:
refactor(compiler): update ordering in template pipeline

Updates the instruction ordering logic in the template pipeline.

crisbeto commented:
Pushed a couple of tests for the property/listener ordering since it ended up being a bit of a gotcha. Also linking the [pipeline-specific TGP](http://test/OCL:603225763:BASE:603229220:1706757332827:1d068782) courtesy of @dylhunn. It has a handful of flakes and one pre-existing failure that is accounted for.

jessicajaniuk commented:
This PR was merged into the repository by commit 3b892e98b00e0e4f74b1e46f1a8d7b8cef0b60a1.
